lib/prism/node_ext.rb

```diff
@@ -83,7 +83,8 @@ def full_name_parts
         current = current.parent
       end
 
-      parts.unshift(current&.name || :"")
+      name = current.is_a?(ConstantReadNode) ? (current&.name || :"") : current.slice
+      parts.unshift(name)
```

Comment:
I don't really want to use `slice` here since it could end up with a whole mess of stuff. For example:

```ruby
foo. # comment
  bar::Baz::Qux
```

You end up with `["foo. # comment\n  bar", :Baz, :Qux]` and `"foo. # comment\n  bar::Baz::Qux"`.

Honestly, I'm not sure of the best solution here, but I think I would prefer to keep the node around in the list and just say it's one of the "parts". That would be:

```suggestion
      part = current.is_a?(ConstantReadNode) ? (current&.name || :"") : current
      parts.unshift(part)
```

Then `full_name` below would need to change as well. I wouldn't want people to call that method and it drop data or be confusing. For example, if you have `var::Foo`, you wouldn't want `full_name` to return `::Foo` (which would be wrong, it's not a top-level constant) and you wouldn't want it to return `Foo` (which is wrong, it should be scoped). There's no winning in this scenario. I think honestly it should raise a special error that indicates that the full name could not be created if the first part is not a symbol.

Reply:
@kddnewton Is this a special case of full_name or a general issue of emitting syntax from the tree?  `syntax` of SelfNode could print out `self`. constant read would be `Foo`.  Constant Path would be syntax of the children.   The nodes know the logic for putting `::` in too.

Reply:
I think you're basically talking about round-tripping back to the original source. That's what I'm trying to solve with https://github.com/ruby-syntax-tree/syntax_tree-prism. This API in particular is being used for indexing code, so it kind of fits into a small niche of nice to haves.